services/configuration/src/transactions.rs

```diff
@@ -0,0 +1,320 @@
+        }
+
+        Ok(())
+    }
+
+    /// Saves this proposal to the service schema.
+    fn save(&self, fork: &mut Fork, cfg: &StoredConfiguration, cfg_hash: Hash) {
+        let prev_cfg = CoreSchema::new(fork.as_ref())
+            .configs()
+            .get(&cfg.previous_cfg_hash)
+            .unwrap();
+
+        // Start writing to storage.
+        // NB. DO NOT write to the service schema anywhere else during `Propose::execute`, it may
+        // break invariants.
+        let mut schema = Schema::new(fork);
+
+        let propose_data = {
+            let mut votes_table = schema.votes_by_config_hash_mut(&cfg_hash);
+            debug_assert!(votes_table.is_empty());
+
+            let num_validators = prev_cfg.validator_keys.len();
+            for _ in 0..num_validators {
+                votes_table.push(MaybeVote::none());
+            }
+
+            ProposeData::new(
+                self.clone(),
+                &votes_table.root_hash(),
+                num_validators as u64,
+            )
+        };
+
+        {
+            let mut propose_data_table = schema.propose_data_by_config_hash_mut();
+            debug_assert!(propose_data_table.get(&cfg_hash).is_none());
+            propose_data_table.put(&cfg_hash, propose_data);
+        }
+
+        schema.config_hash_by_ordinal_mut().push(cfg_hash);
+    }
+}
+
+impl Transaction for Propose {
+    fn verify(&self) -> bool {
+        self.verify_signature(self.from())
+    }
+
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+        let (cfg, cfg_hash) = self.precheck(fork.as_ref()).map_err(|err| {
+            error!("Discarding propose {:?}: {}", self, err);
+            err
+        })?;
+
+        self.save(fork, &cfg, cfg_hash);
+        trace!("Put propose {:?} to config_proposes table", self);
+        Ok(())
+    }
+}
+
+impl Vote {
+    /// Checks context-dependent conditions for a `Vote` transaction.
+    ///
+    /// # Return value
+    ///
+    /// Returns a configuration this vote is for on success, or an error (if any).
+    fn precheck(&self, snapshot: &Snapshot) -> Result<StoredConfiguration, VoteError> {
+        use self::CommonError::*;
+        use self::VoteError::*;
+
+        let following_config = CoreSchema::new(snapshot).following_configuration();
+        if let Some(following) = following_config {
+            Err(AlreadyScheduled(following))?;
+        }
+
+        if let Some(validator_id) = validator_index(snapshot, self.from()) {
+            let vote = Schema::new(snapshot)
+                .votes_by_config_hash(self.cfg_hash())
+                .get(validator_id as u64)
+                .unwrap();
+            if !vote.is_none() {
+                Err(AlreadyVoted)?;
+            }
+        } else {
+            Err(UnknownSender)?;
+        }
+
+        let propose = Schema::new(snapshot).propose(self.cfg_hash()).ok_or_else(
+            || {
+                UnknownConfigRef(*self.cfg_hash())
+            },
```

Comment:
Braces are not necessary?

Reply:
They are added by the formatter, unfortunately.